src/utils/hooks/useGetResponsesAndElementState.tsx

```diff
@@ -0,0 +1,144 @@
+import evaluateExpression from '@openmsupply/expression-evaluator'
+
+const useGetResponsesAndElementState = (props: { serialNumber: string }) => {
+  const { serialNumber } = props
+  const [responsesByCode, setResponsesByCode] = useState({})
+  const [responsesFullByCode, setResponsesFullByCode] = useState({})
+  const [elementsExpressions, setElementsExpressions] = useState([{}])
+  const [elementsState, setElementsState] = useState({})
+  const [error, setError] = useState('')
+  const [loading, setLoading] = useState(true)
+  const { data, loading: apolloLoading, error: apolloError } = useGetElementsAndResponsesQuery({
+    variables: { serial: serialNumber },
+  })
+
+  useEffect(() => {
+    const error = checkForApplicationErrors(data)
+
+    if (error) {
+      setError(error)
+      setLoading(false)
+      return
+    }
+
+    if (apolloError) return
+
+    const applicationResponses = data?.applicationBySerial?.applicationResponses
+      .nodes as ApplicationResponse[]
+
+    const templateSections = data?.applicationBySerial?.template?.templateSections
+      .nodes as TemplateSection[]
+
+    const templateElements: TemplateElementState[] = []
+
+    templateSections.forEach((sectionNode) => {
+      const elementsInSection = sectionNode.templateElementsBySectionId?.nodes as TemplateElement[]
+      elementsInSection.forEach((element) => {
+        templateElements.push(element as TemplateElementState) // No idea why ...[spread] doesn't work here.
+      })
+    })
+
+    const currentResponses = {} as ResponsesByCode
+    const currentFullResponses = {} as ResponsesFullByCode
+
+    applicationResponses.forEach((response) => {
+      const code = response?.templateElement?.code
+      if (code) {
+        currentResponses[code] = response?.value?.text
+        currentFullResponses[code] = response?.value
+      }
+    })
+
+    setResponsesByCode(currentResponses)
+    setResponsesFullByCode(currentFullResponses)
+    setElementsExpressions(templateElements)
+  }, [data, apolloError])
+
+  useEffect(() => {
+    if (responsesByCode)
+      evaluateElementExpressions().then((result) => {
+        setElementsState(result)
+        setLoading(false)
+      })
+  }, [responsesByCode])
+
+  async function evaluateElementExpressions() {
+    const promiseArray: Promise<EvaluatedElement>[] = []
+    elementsExpressions.forEach((element) => {
+      promiseArray.push(evaluateSingleElement(element))
+    })
+    const evaluatedElements = await Promise.all(promiseArray)
+    const elementsState: any = {}
+    evaluatedElements.forEach((element) => {
+      elementsState[element.code] = {
+        id: element.id,
+        category: element.category,
+        isEditable: element.isEditable,
+        isRequired: element.isRequired,
+        isVisible: element.isVisible,
+        // isValid: element.isValid,
+      }
+    })
+    return elementsState
+  }
+
+  async function evaluateSingleElement(element: any): Promise<EvaluatedElement> {
```

Comment:
```suggestion
  async function evaluateSingleElement(element: TemplateElementState): Promise<EvaluatedElement> {
```

Reply:
I'm also thinking how we could combine these three types: `TemplateElementState`, `EvaluatedElement` and (a new one I have created to be used as following (for other PRs):
```
interface ElementState {
  id: number
  title: string
  elementTypePluginCode: string
  section: number
  category: TemplateElementCategory
  isRequired: boolean
  isVisible: boolean
  isValid: boolean
  // validationMessage: string | null
  isEditable: boolean
}

interface ApplicationElementState {
  [key: string]: ElementState
}
```

Since these 3 have so much in common it would be good to have some intersection between types. Can you think on something? I hasn't spent to much time on it, just think it was kind of annoying to have to add fields in these 3 while I was extending this to use somewhere else that required more fields...

Reply:
Okay, have done the Typing on the suggestion.

I'm not sure how to combine these, maybe I'll wait to see your PR and then we'll consider all three of them in its own issue?

Reply:
Yes, all right. I think it would be better to merge this one already. And then we will tidy up these types altogether in the other PR. I was trying to avoid to many changes to the other one, but it will be alright to wait until then to change this.